### DISCUSSION TITLE
Best practices for FastAPI projects with SQLModel ORM and PostgreSQL database

### DISCUSSION THREAD

**Discussion by rizwan-ai** [fastapi/fastapi#9936] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
async def create_db():
    """
    Create database
    """
    async with DB_ENGINE.begin() as conn:
        # await conn.run_sync(SQLModel.metadata.drop_all)
        await conn.run_sync(SQLModel.metadata.create_all)

if __name__ == "__main__":
    asyncio.run(create_db())
```

### Description

What is the recommended Python development workflow for FastAPI projects?
Should we use venv, pipenv, or poetry for managing dependencies?

When using FastAPI, which ORM provides the best integration and performance?
Is it preferable to use SQLAlchemy ORM or SQLModel ORM?

FastAPI with SQLModel currently supports SQLAlchemy version SQLAlchemy==1.4.41. Why doesn't it support the latest version?
What could be the reasons for the limited support, and should we expect updates for the latest SQLAlchemy version?

What is the best practice for using FastAPI with SQLModel and databases?
Should we use psycopg2, asyncpg, or psycopg (psycopg 3.x) for interacting with PostgreSQL databases in this setup?

Seeking advice on the ideal project structure for a FastAPI project with SQLModel ORM, PostgreSQL (asyncpg), Alembic, and MinIO.

### Operating System

Linux

### FastAPI Version

0.100.0

### Python Version

3.11.4

**Comment by nameer** [comment]
I believe most of your questions will be answered with [this repository](https://github.com/tiangolo/full-stack-fastapi-postgresql).

**Reply by rizwan-ai** [reply]
3 years old repository

**Comment by jgould22** [comment, marked as answer]
>What is the recommended Python development workflow for FastAPI projects? Should we use venv, pipenv, or poetry for managing dependencies?

You should use whichever you are most familiar and comfortable with. I would use poetry but only because I know it and generally use it.

> When using FastAPI, which ORM provides the best integration and performance? 

SQLModel is simply pydantic types integrated with  SQLAlchemy models. It isnt really so much an ORM as a companion library to SQLA and Pydantic to streamline some of the boiler plate you might require converting between these two types

> Is it preferable to use SQLAlchemy ORM or SQLModel ORM?

The FastAPI developer would tell you to use SQLModel as he wrote it and obviously thought it was a good improvement. I personally stick to SQLAlchemy, I am not to fussed with keeping my models and schemas in sync and did not feel the need to add another abstraction on top of SQLA.

> FastAPI with SQLModel currently supports SQLAlchemy version SQLAlchemy==1.4.41. Why doesn't it support the latest version?

It depends on a single developer who prefers to maintain control of his repositories. It doesnt support the latest version because he has not yet updated it. The bus factor of any single library you intend to use should be a consideration before you choose to use the library (including fastapi).

> What could be the reasons for the limited support, and should we expect updates for the latest SQLAlchemy version?

I would expect updates whenever the dev gets around to, he has a lot of projects though and I am not sure how he prioritizes the work.  

> What is the best practice for using FastAPI with SQLModel and databases?

That is probably better asked in the SQLModel repo.

> Should we use psycopg2, asyncpg, or psycopg (psycopg 3.x) for interacting with PostgreSQL databases in this setup?

If  SQLModel is on SQLA 1.4.x and that version of SQLA does not support psycopg 3 I would suspect you should use asyncpg if you are using an async session or psycopg2 if you are using a sync session.

> Seeking advice on the ideal project structure for a FastAPI project with SQLModel ORM, PostgreSQL (asyncpg), Alembic, and MinIO.

This is an entirely subjective thing. I lay mine out sort of like the below but I don't tend to have frontend components in my projects 
The linked cookie cutter is old but the layout is still relevant  
```
- docker
- alembic
- my_app
   - main.py
   - core
       - settings.py
       - logging.py
       - metrics.py 
   - lib
   - schema
   - models
   - databases
   - api
- README.md
- pyproject.py
```

**Reply by rizwan-ai** [reply]
Thanks @jgould22

Pipenv creates Pipfile and Pipfile.lock, when installing a new package then pipenv update all existing dependencies,
like poetry also the same work

**Comment by mofalih64** [comment]
I stumbled on this article months ago, and I found it super useful , hope the same to you.
[https://praciano.com.br/fastapi-and-async-sqlalchemy-20-with-pytest-done-right.html](https://praciano.com.br/fastapi-and-async-sqlalchemy-20-with-pytest-done-right.html)
